File: ts/src/rule-id.ts
import type { RuleFunc, Specs } from "./types.js";
import { funcID } from "./func-id.js";

export function ruleID(fn: RuleFunc, specs: Specs): string {
  const id = funcID(fn);
  if (specs.length === 0) return id;
  const names = specs.map(s => JSON.stringify(s)).sort();
  return id + "#" + names.join("+");
}

File: py/src/rulecat/rule_id.py
from __future__ import annotations
import json
from typing import Any

from rulecat.func_id import func_id


def rule_id(fn: Any, specs: list[Any]) -> str:
    fid = func_id(fn)
    if not specs:
        return fid
    names = sorted(json.dumps(s.__dict__, default=str) for s in specs)
    return fid + "#" + "+".join(names)
